pkg/virtual/apiexport/authorizer/maximal_permission_policy.go

```diff
@@ -0,0 +1,257 @@
+	getAPIBindingReferenceForAttributes func(attr authorizer.Attributes, clusterName logicalcluster.Name) (ref *apisv1alpha1.ExportReference, found bool, err error)
+}
+
+// NewMaximalPermissionAuthorizer creates an authorizer that checks the maximal permission policy
+// for the requested resource if the resource is a claimed resource in the requested API export.
+// The check is omitted if the requested resource itself is not associated with another API export.
+//
+// If the request is a cluster request the authorizer skips authorization if the request is not for a bounded resource.
+// If the request is a wildcard request this check is skipped because no unique API binding can be determined.
+func NewMaximalPermissionAuthorizer(deepSARClient kubernetes.ClusterInterface, apiExportInformer apisinformers.APIExportInformer, apiBindingInformer apisinformers.APIBindingInformer) authorizer.Authorizer {
+	apiExportLister := apiExportInformer.Lister()
+	apiExportIndexer := apiExportInformer.Informer().GetIndexer()
+	apiBindingIndexer := apiBindingInformer.Informer().GetIndexer()
+	indexers.AddIfNotPresentOrDie(apiBindingIndexer, indexers.ClusterScoped())
+	indexers.AddIfNotPresentOrDie(apiExportIndexer, indexers.ClusterScoped())
+
+	return &maximalPermissionAuthorizer{
+		getAPIExport: func(clusterName, apiExportName string) (*apisv1alpha1.APIExport, error) {
+			return apiExportLister.Get(clusters.ToClusterAwareKey(logicalcluster.New(clusterName), apiExportName))
+		},
+		getAPIExportByIdentity: func(identityHash string) (*apisv1alpha1.APIExport, error) {
+			exports, err := apiExportIndexer.ByIndex(indexers.APIExportByIdentity, identityHash)
+			if err != nil {
+				return nil, err
+			}
+			if len(exports) == 0 {
+				return nil, errors.New("export not found")
+			}
+			return exports[0].(*apisv1alpha1.APIExport), nil
+		},
+		newDeepSARAuthorizer: func(clusterName logicalcluster.Name) (authorizer.Authorizer, error) {
+			return delegated.NewDelegatedAuthorizer(clusterName, deepSARClient)
+		},
+		getAPIBindingReferenceForAttributes: func(attr authorizer.Attributes, clusterName logicalcluster.Name) (*apisv1alpha1.ExportReference, bool, error) {
+			return kcpauth.GetAPIBindingReferenceForAttributes(apiBindingIndexer, attr, clusterName)
+		}}
+}
+
+func (a *maximalPermissionAuthorizer) Authorize(ctx context.Context, attr authorizer.Attributes) (authorizer.Decision, string, error) {
+	lcluster, err := genericapirequest.ClusterNameFrom(ctx)
+	if err != nil {
+		kaudit.AddAuditAnnotations(
+			ctx,
+			VirtualAPIExportMaximumPermissionPolicyAuditDecision, kcpauth.DecisionNoOpinion,
+			VirtualAPIExportMaximumPermissionPolicyAuditReason, fmt.Sprintf("error getting cluster from request: %v", err),
+		)
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, err
+	}
+
+	if lcluster != logicalcluster.Wildcard {
+		_, bound, err := a.getAPIBindingReferenceForAttributes(attr, lcluster)
```

Comment:
I'm not sure I follow this logic - why is checking to see if the API is bound part of authz here?

Reply:
Two cases:
- wildcard request: no unique API binding can be determined => this case is skipped
- cluster request: a unique API binding can be determined => according to @sttts we should only forward authz iff the API binding is bound, it resembles the same logic as in the API binding authorizer

Maybe omitting this here is good enough nevertheless (just follow the "wildcard request" case) @sttts ?

Reply:
It was wrong indeed, I removed it. If the request is targetting a concrete cluster (not to be confused with the api export cluster being present of the virtual apiserver URL) we could go the route of checking API bindings as well, however after a paring session with @sttts we figured that this authorizer should focus on the API exports in the first place.